// sanity-plugin-mux-input/build/components/Video.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _hls = _interopRequireDefault(require("hls.js"));

var _bar = _interopRequireDefault(require("part:@sanity/components/progress/bar"));

var _default2 = _interopRequireDefault(require("part:@sanity/components/buttons/default"));

var _assets = require("../actions/assets");

var _secrets = require("../actions/secrets");

var _getPosterSrc = _interopRequireDefault(require("../util/getPosterSrc"));

var _getVideoSrc = _interopRequireDefault(require("../util/getVideoSrc"));

var _Video = _interopRequireDefault(require("./Video.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var NOOP = () => {
  /* intentional noop */
};

var propTypes = {
  assetDocument: _propTypes.default.object.isRequired,
  autoload: _propTypes.default.bool,
  onCancel: _propTypes.default.func,
  onReady: _propTypes.default.func
};

class MuxVideo extends _react.Component {
  constructor(props) {
    super(props);

    _defineProperty(this, "videoContainer", /*#__PURE__*/_react.default.createRef());

    _defineProperty(this, "hls", null);

    _defineProperty(this, "handleVideoClick", event => {
      this.setState({
        showControls: true
      });
      this.hls.startLoad(0);

      if (this.props.onReady) {
        this.props.onReady(event);
      }
    });

    _defineProperty(this, "handleCancelButtonClicked", event => {
      if (this.props.onCancel) {
        this.props.onCancel(event);
      }
    });

    this.state = {
      posterUrl: null,
      source: null,
      isLoading: true,
      error: null,
      isDeletedOnMux: false,
      secrets: null
    };
  } // eslint-disable-next-line complexity


  static getDerivedStateFromProps(nextProps) {
    var isLoading = true;
    var assetDocument = nextProps.assetDocument;

    if (assetDocument && assetDocument.status === 'preparing') {
      isLoading = 'Preparing the video';
    }

    if (assetDocument && assetDocument.status === 'waiting_for_upload') {
      isLoading = 'Waiting for upload to start';
    }

    if (assetDocument && assetDocument.status === 'waiting') {
      isLoading = 'Processing upload';
    }

    if (assetDocument && assetDocument.status === 'ready') {
      isLoading = false;
    }

    if (assetDocument && typeof assetDocument.status === 'undefined') {
      isLoading = false;
    }

    return {
      isLoading
    };
  }

  componentDidMount() {
    this.video = /*#__PURE__*/_react.default.createRef();
    this.setState(MuxVideo.getDerivedStateFromProps(this.props));
    (0, _secrets.fetchSecrets)().then((_ref) => {
      var secrets = _ref.secrets;
      return this.setState({
        secrets
      });
    });
  }

  componentDidUpdate(prevProps, prevState) {
    if (!this.state.isLoading && this.state.secrets && this.state.source === null) {
      this.resolveSourceAndPoster(this.props.assetDocument);
    }

    if (this.state.source !== null && this.video.current && !this.video.current.src) {
      // eslint-disable-next-line react/no-did-update-set-state
      this.setState({
        error: null
      });
      this.attachVideo();
    }

    if (this.state.source !== null && this.state.source !== prevState.source) {
      // eslint-disable-next-line react/no-did-update-set-state
      this.setState({
        error: null,
        showControls: false
      });

      if (this.hls) {
        this.hls.destroy();
      }

      this.attachVideo();
    }
  }

  resolveSourceAndPoster(assetDocument) {
    var playbackId = assetDocument.playbackId;
    var options = {
      isSigned: assetDocument.data.playback_ids[0].policy === 'signed',
      signingKeyId: this.state.secrets.signingKeyId || null,
      signingKeyPrivate: this.state.secrets.signingKeyPrivate || null
    };
    var source = (0, _getVideoSrc.default)(playbackId, options);
    var posterUrl = (0, _getPosterSrc.default)(playbackId, options);
    this.setState({
      source,
      posterUrl
    });
  }

  getVideoElement() {
    return this.video && this.video.current;
  }

  attachVideo() {
    var _this$props = this.props,
        assetDocument = _this$props.assetDocument,
        autoload = _this$props.autoload;

    if (_hls.default.isSupported()) {
      this.hls = new _hls.default({
        autoStartLoad: autoload
      });
      this.hls.loadSource(this.state.source);
      this.hls.attachMedia(this.video.current);
      this.hls.on(_hls.default.Events.MANIFEST_PARSED, () => {
        if (this.videoContainer.current) {
          this.videoContainer.current.style.display = 'block';
        }

        if (this.props.onReady) {
          this.props.onReady();
        }
      });
      this.hls.on(_hls.default.Events.ERROR, (event, data) => {
        switch (data.type) {
          case _hls.default.ErrorTypes.NETWORK_ERROR:
            if (this.videoContainer.current) {
              this.videoContainer.current.style.display = 'none';
            }

            this.setState({
              error: data
            });
            (0, _assets.getAsset)(assetDocument.assetId).then(response => {
              this.setState({
                isDeletedOnMux: false
              });
            }).catch(err => {
              if (err.message.match(/404/)) {
                this.setState({
                  isDeletedOnMux: true
                });
                return;
              }

              console.error(data, err); // eslint-disable-line no-console
            });
            break;

          default:
            console.error(data);
          // eslint-disable-line no-console
        }
      });
    } else if (this.video.current.canPlayType('application/vnd.apple.mpegurl')) {
      this.video.current.src = this.state.source;
      this.video.current.addEventListener('loadedmetadata', () => {
        this.hls.loadSource(this.state.source);
        this.hls.attachMedia(this.video.current);
      });
    }
  }

  // eslint-disable-next-line complexity
  render() {
    var _this$state = this.state,
        posterUrl = _this$state.posterUrl,
        isLoading = _this$state.isLoading,
        error = _this$state.error;
    var _this$props2 = this.props,
        assetDocument = _this$props2.assetDocument,
        autoload = _this$props2.autoload;

    if (!assetDocument || !assetDocument.status) {
      return null;
    }

    if (isLoading) {
      return /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("div", {
        className: _Video.default.progressBar
      }, /*#__PURE__*/_react.default.createElement(_bar.default, {
        percent: 100,
        text: isLoading !== true && isLoading || 'Waiting for MUX to complete the file',
        isInProgress: true,
        showPercent: true,
        animation: true,
        color: "primary"
      })), /*#__PURE__*/_react.default.createElement("div", {
        className: _Video.default.uploadCancelButton
      }, /*#__PURE__*/_react.default.createElement(_default2.default, {
        onClick: this.handleCancelButtonClicked
      }, "Cancel")));
    }

    var showControls = autoload || this.state.showControls;
    return /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("div", {
      ref: this.videoContainer,
      className: _Video.default.videoContainer
    }, /*#__PURE__*/_react.default.createElement("video", {
      className: _Video.default.root,
      onClick: autoload ? NOOP : this.handleVideoClick,
      controls: showControls,
      ref: this.video,
      poster: posterUrl
    })), error && /*#__PURE__*/_react.default.createElement("div", {
      className: [_Video.default.videoContainer, _Video.default.videoError].join(' ')
    }, /*#__PURE__*/_react.default.createElement("p", null, "There was an error loading this video (", error.type, ")."), this.state.isDeletedOnMux && /*#__PURE__*/_react.default.createElement("p", null, /*#__PURE__*/_react.default.createElement("strong", null, "The video is deleted on MUX.com"))));
  }

}

_defineProperty(MuxVideo, "defaultProps", {
  autoload: true
});

MuxVideo.propTypes = propTypes;
var _default = MuxVideo;
exports.default = _default;
